src/bun.js/node/win_watcher.zig

```diff
@@ -0,0 +1,391 @@
+const std = @import("std");
+const bun = @import("root").bun;
+const uv = bun.windows.libuv;
+const Path = @import("../../resolver/resolve_path.zig");
+const Fs = @import("../../fs.zig");
+const Mutex = @import("../../lock.zig").Lock;
+const string = bun.string;
+const JSC = bun.JSC;
+const VirtualMachine = JSC.VirtualMachine;
+const StoredFileDescriptorType = bun.StoredFileDescriptorType;
+const Output = bun.Output;
+
+var default_manager: ?*WinPathWatcherManager = null;
+
+// TODO: make this a generic so we can reuse code with path_watcher
+// TODO: we probably should use native instead of libuv abstraction here for better performance
+pub const WinPathWatcherManager = struct {
+    const GenericWatcher = @import("../../watcher.zig");
+    const options = @import("../../options.zig");
+    pub const Watcher = GenericWatcher.NewWatcher(*WinPathWatcherManager);
+    const log = Output.scoped(.WinPathWatcherManager, false);
+    main_watcher: *Watcher,
+
+    watchers: bun.BabyList(?*PathWatcher) = .{},
+    watcher_count: u32 = 0,
+    vm: *JSC.VirtualMachine,
+    file_paths: bun.StringHashMap(PathInfo),
+    deinit_on_last_watcher: bool = false,
+    const PathInfo = struct {
+        fd: StoredFileDescriptorType = .zero,
+        is_file: bool = true,
+        path: [:0]const u8,
+        dirname: string,
+        refs: u32 = 0,
+        hash: Watcher.HashType,
+    };
+
+    fn _fdFromAbsolutePathZ(
+        this: *WinPathWatcherManager,
+        path: [:0]const u8,
+    ) !PathInfo {
+        if (this.file_paths.getEntry(path)) |entry| {
+            var info = entry.value_ptr;
+            info.refs += 1;
+            return info.*;
+        }
+        const cloned_path = try bun.default_allocator.dupeZ(u8, path);
+        errdefer bun.default_allocator.free(cloned_path);
+
+        if (std.fs.openDirAbsoluteZ(cloned_path, .{
```

Comment:
Can we use bun's fs open instead? This will potentially segfault on Windows

Reply:
fixed in https://github.com/oven-sh/bun/pull/8379/commits/b323bc66cfc1bb68130489c1f03e90679f6667a2